client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -250,12 +250,32 @@ public void flush() {
                         // available writers.
                         success = false;
                         log.warn("Flush on segment {} failed due to {}, it will be retried.", writer.getSegmentName(), e.getMessage());
+                        tryWaitForSuccessors();
                         break;
                     }
                 }
             }
         }
     }
+    
+    /**
+     * This is used by flush to optimistically wait for the `handleLogSealed` work to be completed
+     * to prevent a busy loop while waiting for the segments to be updated following a sealed
+     * segment. Please note there are no guarantees about when this method returns. It can be
+     * thought of as a sleep.
+     */
+    @GuardedBy("writeFlushLock")
+    private void tryWaitForSuccessors() {
+        while (!sealedSegmentQueue.isEmpty()) {
+            // A background thread should be waking up to process things. Give it a moment.
+            Exceptions.handleInterrupted(() -> Thread.sleep(10));
+        }
+        synchronized (writeSealLock) {
```

Comment:
No new writes are allowed once `writeFlushLock` is held by flush(),this implies that writer need not try to acquire the `writeSealLock ` to verify that retransmit to sealed segments are complete, right ?

Also, `sealedSegmentQueue.isEmpty()` check would suffice in the scenario where retransmit of inflight events face a sealed segment exception..

Reply:
We want to make sure that when `flush` is called it does not return until all the segments have been flushed. It aquires the `writeFlushLock` to prevent future writes. However there may be a concurrent seal. We just need to keep retrying calling flush on the writers until all have flushed. The previous logic was "correct" however in the event there was an ongoing seal it was wasteful because it would keep trying to flush a sealed segment over and over. The solution here it to attempt to block on the seal being completed. `sealedSegmentQueue.isEmpty()` Would tell us that no seal operation is pending. This is a necessary but not a sufficient criteria, because there could be still one in progress. The `writeSealLock` is not needed to do anything here, (Notice nothing is done inside of the lock) it's just done to queue the flush thread on an ongoing seal if one is indeed in progress.

Reply:
ok 👍